app/assets/javascripts/components/overview/campaign_button.jsx

```diff
@@ -1,25 +1,109 @@
 import React from 'react';
+import createReactClass from 'create-react-class';
 import PropTypes from 'prop-types';
-import PopoverButton from '../high_order/popover_button.jsx';
-import CampaignStore from '../../stores/campaign_store.js';
+import { connect } from 'react-redux';
+import Select from 'react-select';
+import _ from 'lodash';
 
-const campaignIsNew = campaign => CampaignStore.getFiltered({ title: campaign }).length === 0;
+import { getAvailableCampaigns } from '../../selectors';
+
+import PopoverExpandable from '../high_order/popover_expandable.jsx';
+import Conditional from '../high_order/conditional.jsx';
+import CourseUtils from '../../utils/course_utils.js';
+
+import { removeCampaign, fetchAllCampaigns, addCampaign } from '../../actions/campaign_actions';
+
+const CampaignButton = createReactClass({
+  displayName: 'CampaignButton',
+
+  getKey() {
+    return `add_campaign`;
+  },
+
+  PropTypes: {
+    campaigns: PropTypes.array,
+    allCampaigns: PropTypes.array
+  },
+
+  handleChangeCampaign(val) {
+    return this.props.addCampaign(this.props.course_id, val.value);
+  },
+
+  removeCampaign(campaignId) {
+    this.props.removeCampaign(this.props.course_id, campaignId);
+  },
+
+  addCampaign(campaignId) {
+    this.props.addCampaign(this.props.course_id, campaignId);
+  },
+
+  render() {
+    let campaignList;
+    let campaignSelect;
+    let campaignClass;
+    const campaignOptions = this.props.allCampaigns.map(campaign => {
+      return { label: campaign, value: campaign };
+    });
+    if (this.props.editable) {
```

Comment:
One possible way to handle this would be to split this into two display components, one for 'edit' mode and one for 'display' mode. Then the 'fetchAllCampaigns', if it hasn't been done yet, could be done when the 'edit' mode version mounts. This nested if/else structure is also likely to be confusing for future readers of this file, and breaking it into simpler display components would make the possible render paths more clear.